File: commons/ocaml/answer.ml
open Core

let timer (f : unit -> unit) =
  let start_time = Time_ns.now () in
  f ();
  let end_time = Time_ns.now () in
  let elapsed = Time_ns.diff end_time start_time in
  Printf.printf "Runtime (ns): %d\n" (Time_ns.Span.to_int_ns elapsed)

let part (n : int) (expected : string) (actual : string) =
  match String.equal expected actual with
  | true -> Printf.printf "Part %d: %s\n" n actual
  | false ->
      let error_message =
        Printf.sprintf "Part %d: expected %s found %s" n expected actual
      in
      raise (Invalid_argument error_message)

let part1 (expected : 'a) (actual : 'a) (f : 'a -> string) =
  part 1 (f expected) (f actual)

let part2 (expected : 'a) (actual : 'a) (f : 'a -> string) =
  part 2 (f expected) (f actual)

File: commons/elixir/answer.ex
defmodule Answer do
  def timer(f) do
    start_time = System.monotonic_time(:nanosecond)
    f.()
    end_time = System.monotonic_time(:nanosecond)
    IO.puts("Runtime (ns): #{end_time - start_time}")
  end

  def part1(expected, actual) do
    part(1, expected, actual)
  end

  def part2(expected, actual) do
    part(2, expected, actual)
  end

  defp part(n, expected, actual) do
    if expected == actual do
      IO.puts("Part #{n}: #{actual}")
    else
      raise RuntimeError, "Part #{n}: expected #{expected} found #{actual}"
    end
  end
end
